File: python3/rotate_array.py
# leetcode

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """

        leng = len(nums)
        k = k % leng
        
        pref_list = nums[-k:]
        pref = len(pref_list)
        suf_list = nums[:-k]
        suf = len(suf_list)
        
        # pref
        for idx in range(pref):
            nums[idx] = pref_list[idx]
            
        # suf
        for idx in range(suf):
            nums[idx+pref] = suf_list[idx]

File: golang/rotate_array.go
package golang

// Space	: O(k)
// Time	: O(n)

func rotate(nums []int, k int) {
	n := len(nums)
	k %= n
	var temp []int

	for i := 0; i < k; i++ {
		temp = append(temp, nums[n-k+i])
	}

	for j := n - 1; j >= 0; j-- {
		if j >= k {
			nums[j] = nums[j-k]
		} else {
			nums[j] = temp[j]
		}

	}
}
